Fix subtle bug in listify function + simplify list munging

jdavcs opened the pull request:
Subtle bug (discovered while upgrading SQLAlchemy syntax.)

The `ModelManager._munge_filters` method is supposed to "combine two lists into a single list" (as per docstring; which is also the right behavior). However, if one of the arguments is `None` and the other argument is not a list, it will return the other argument unchanged. This becomes a problem when one tries to iterate over the returned value, which is not an unlikely scenario:
```py
filters = self._munge_filters(my_new_filter, current_filters)  # current_filters is None
stmt = select(Foo).where(*filters)  # this will break
```

The obvious fix is to replace the existing implementation with a one-liner reusing our existing `listify` util function on both arguments: `return listify(listA) + listify(listB)`. However, this exposed a bug: if one of the arguments is an object with the special `__bool__` method overridden, `listify` may return an empty list - which is the case with SQLAlchemy filters. A filter is a sql expression (e.g. `Foo.bar == condition`); so checking `if not my_filter` will simply apply `not` to the sql expression and negate it, returning `True` -  and that 's where the code assumes the item as `None` and returns an empty list. Explicitly checking for `None` fixes this.

List munging moved to util.

Unit tests included.




## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Add new munge_lists function + unit test

jdavcs committed:
Replace ModelManager.munge_lists with new util function

Also, drop _munge_filters and use util function directly.

mvdbeek commented on lib/galaxy/util/__init__.py:
```diff
@@ -1069,7 +1069,7 @@ def listify(item: Any, do_strip: bool = False) -> List:
     :returns:           The input as a list
     """
-    if not item:
+    if item is None or item == "":
```
This seems like a pretty risky change. Can you move the special casing to munge_filters ? It does change things i.e. when item is 0.

nsoranzo replied:
It surely is counter-intuitive that `listify(0)` and `listify(False)` return `[]`, but as Marius says we cannot change the behaviour of this function any more.
We probably need a `listify_new` function that does what you were proposing here, and use that in new code / when refactoring.

mvdbeek commented on lib/galaxy/util/__init__.py:
```diff
@@ -1882,3 +1882,14 @@ def enum_values(enum_class):
     return [value.value for value in enum_class.__members__.values()]
+
+
+def munge_lists(listA: Any, listB: Any) -> List:
```
I don't think we want more functions in `util/__init__.py`, which we have to maintain indefinitely as the public api. Apart from that `munge_lists` is not a good name and we're not re-using this anywhere.

mvdbeek replied:
@jmchilton can you elaborate ?

jdavcs committed:
Do not reuse util.listify

Move function to managers.base; rename.
Do not reuse listify; implement standalone function.
Add test case for tuples.
(See PR code review for discussion)

mvdbeek reviewed (approved):
Thank you, I think this looks good.

jdavcs committed:
Do not make a list out of a tuple

Because the passed argument is not a SA filter, but our own parsed
filter, which is a tuple.
